File: python/mirage/core/wandb/stat.py
from mirage.accessor.wandb import WandbAccessor
from mirage.cache.index import NULL_INDEX, IndexCacheStore, RAMIndexCacheStore
from mirage.core.hierarchy.probe import assert_listed, resolve_entry
from mirage.core.wandb.pathing import parts
from mirage.core.wandb.readdir import readdir
from mirage.types import ContentType, FileStat, FileType, PathSpec
from mirage.utils.errors import enoent


async def stat(accessor: WandbAccessor,
               path: PathSpec,
               index: IndexCacheStore = NULL_INDEX) -> FileStat:
    ps = parts(accessor, path)
    if not ps:
        return FileStat(name="/", type=FileType.DIRECTORY, size=0)
    if index is NULL_INDEX or index is None:
        index = RAMIndexCacheStore()
    await assert_listed(readdir, accessor, path, index)
    found = await resolve_entry(readdir, accessor, path, index)
    if found is None:
        raise enoent(path)
    directory = found.resource_type == "wandb/directory"
    content = None
    if not directory:
        if len(ps) > 4:
            content = ContentType.BINARY
        elif ps[-1].endswith(".json"):
            content = ContentType.JSON
        else:
            content = ContentType.TEXT
    return FileStat(name=ps[-1],
                    type=FileType.DIRECTORY if directory else FileType.FILE,
                    size=found.size,
                    content=content)

File: typescript/packages/core/src/core/wandb/stat.ts
import type { WandbAccessor } from '../../accessor/wandb.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import { RAMIndexCacheStore } from '../../cache/index/ram.ts'
import { ContentType, FileStat, FileType, type PathSpec } from '../../types.ts'
import { enoent } from '../../utils/errors.ts'
import { assertListed, resolveEntry } from '../hierarchy/probe.ts'
import { parts } from './pathing.ts'
import { readdir } from './readdir.ts'
export async function stat(
  accessor: WandbAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<FileStat> {
  const ps = parts(accessor, path)
  if (ps.length === 0) return new FileStat({ name: '/', type: FileType.DIRECTORY, size: 0 })
  index ??= new RAMIndexCacheStore()
  await assertListed(readdir, accessor, path, index)
  const found = await resolveEntry(readdir, accessor, path, index)
  if (!found) throw enoent(path)
  const name = ps.at(-1) ?? ''
  const directory = found.resourceType === 'wandb/directory'
  let content: ContentType | null = null
  if (!directory) {
    if (ps.length > 4) content = ContentType.BINARY
    else if (name.endsWith('.json')) content = ContentType.JSON
    else content = ContentType.TEXT
  }
  return new FileStat({
    name,
    type: directory ? FileType.DIRECTORY : FileType.FILE,
    size: found.size,
    content,
  })
}
